pyproject.toml

```diff
@@ -27,7 +27,7 @@ markers = [
 [tool.ruff]
 # Enable Pyflakes `E` and `F` codes by default.
 select = [
-    "C40",
+    "C4",
```

Comment:
After gh-24935 we should go the other way and make the flake8 rules more selective. It is bad to have a situation where upstream updates break CI in a way that is otherwise perfectly avoidable.

If flake8-comprehensions can add breaking lint rules then we should not use it in a way that is open-ended with respect to new lint rules.

Reply:
@oscarbenjamin Actually, no new lint rule as added by flake8-comprehensions, an existing lint rule C416 was expanded.

Reply:
@oscarbenjamin The proper solution is to pin dependencies and have dependabot or something open PRs to update the dependencies on a periodic basis/

Reply:
No, pinning dependencies is not the proper solution because then we don't get bugfix and other updates automatically. Dependabot is just a whole other waste of time that we don't need generating millions of PRs for mostly irrelevant updates.

The proper solution is that flake8-comprehensions should provide a way to be used that does not mean breakage in new releases.

> an existing lint rule C416 was expanded.

That's not good. The point of listing these rules with codes is so that downstream projects can control what rules are being applied. If flake8-comprehensions is not stable then maybe it shouldn't be used in CI.

Reply:
@oscarbenjamin This is the first non-false positive breaking change or rule addition for flake8-comprehension in years (at least since 2019/2020), its about as stable as flake8 linters are.

Reply:
I agree it probably should have been added a new rule ID, but unfortunately the previous ID already shared an ID for sets and list comprehensions, which was against conventions, but already merged unfortunately.

Reply:
Well if we assume that they will respect backwards compatibility with the codes in future then I still think that we need to have an explicit list of which codes are being used rather than something open ended. There is no benefit in having an open-ended code spec unless we want a new rule to be applied unexpectedly in future (and we don't want that).

Reply:
@oscarbenjamin Maintaining to separate long lists of code seem excessive. The standard process across all the projects I have had with lint support is having the versions pinned and running periodic autoupdates with a pre-commit or dependabot. Even bug fixes could change behavior in the future by removing false negatives in existing codepaths etc, dropping support for the Python version we are on, etc. Flake8 also can and has changed behavior or added new error codes between versions as well and I wouldn't advocate enumerating every rule for it.

Reply:
@oscarbenjamin This is also counter to `ruffs'` config recommendations (see here: https://beta.ruff.rs/docs/configuration/#using-pyprojecttoml )

Reply:
@oscarbenjamin Exactly, dependabot will post that there has been an update and then if a reviewer wants the update to go through they will either need to explicitly add the check to the "IGNORE" list or fix it. Otherwise, new useful checks may not be added even if the codebase is already following that contention (which would allow a later PR to be added that would then violate the convention the codebase is previously following).

For reference @oscarbenjamin, ruff has over 500 rules right now and I'd say at least 100-200+ of them could be applied right now if we wanted to without any functional changes to our codebase (since we already following the proper convention).

Reply:
@oscarbenjamin Furthermore, adding all the codes explicitly to this will not stop bugfixes from finding new violations in existing error codes as what happened with flake8-comprehensions.

Also I understand the concern about PR spam, but you can configure how often the PRs are opened.

For instance for pybind11, we have pre-comimt only updates the linters once a month automatically which works well. 12 automated PRs (at most) a year seem perfectly reasonable to me.

Reply:
IMHO dependabot is an extremely annoying spam producer, I've yet to see it offer any value in any project I'm involved in. I don't think we should add that. I agree that we should explicitly select what linting rules we enforce for a PR to be merged. This should be a minimal starting set (as we've had) and PR by PR new ones can be added if the sympy devs want to add them.

Reply:
@oscarbenjamin I can prototype the pre-commit CI autoupdate bot / process, but in the mean time can this PR be accepted since it just brings ruff better inline with flake8.

Reply:
One thing that would be better than dependabot potentially is a version of it that could update many things in a single PR.

Reply:
I'm not sure I understand the difference. The main thing that concerns me about dependabot et al is minimising the noise from the version update PRs. With dependabot it only updates one thing at a time but a better option would be to have something that just updates everything in one go say once a month if it's possible to do that.

Reply:
That's only updating the precommit hooks though. If we want to go with the idea of pinning versions to prevent CI from breaking then we need to pin all sorts of things like the version of sphinx used in the docs build, the version of numpy, scipy in the optional dependency tests, mypy version for the mypy job etc.

Reply:
@oscarbenjamin I am only advocating pinning linters/formatters (flake8 + ruff). Other dependenceis could be pinned too, but I only was really advocating for pinning the hooks.